### PR TITLE
Exploration of NPM upgrades

### PR DESCRIPTION
**Pull request by sawyerpollard**
Recently, I decided to learn more about Webpack, NPM, and various frontend build tools—and catalogued some of my findings for myself. I ended up updating packages and tweaking config options to learn more about the build system of the Note Editor. I figured I might as well share some of what I did in a draft PR if any of it is useful.

## NPM Updates
The initial big thing I did is that I updated every NPM dependency to the latest version, including to Webpack 5, which introduced a lot of changes. After learning about NPM dependency resolution and peerDependencies, I was able to get rid of the `npm install` error I had been experiencing and the need to use `--legacy-peer-deps`. Additionally, these updates seemed to get the vulnerability count to 0, which probably isn't relevant but cool. There are also now only 2 dependencies listed as deprecated.

I also added a couple NPM scripts for more build targets.

In updating packages, I realized that react-virtualized has a peerDependency requirement that should be updated to handle newer versions of React. Additionally, I discovered that postcss-rtl is unmaintained (https://github.com/vkalinichev/postcss-rtl) and that it is recommended to use postcss-rtlcss (https://github.com/elchininet/postcss-rtlcss). I have added this plugin to PostCSS but have not tested it.

## CSS Tweaks
Another goal was to get rid of any build errors. I was thankfully able to do this by adding `calc()` functions where necessary in CSS.

## Babel Changes
I moved the Babel configuration to its own file, instead of Webpack. This allows it to more easily integrate with ESLint and remedied a missing config file specified in `.eslintrc` with `"configFile": "./.babelrc-es"`. I removed https://babeljs.io/docs/en/babel-plugin-proposal-class-properties, which seems like its included in the latest Babel version. After inspecting the Babel docs, I believe the config is adequate. Additionally, this seems to have fixed ESLint on my computer.

## HtmlWebpackPlugin
I decided to experiment with HtmlWebpackPlugin, which automates the linking of CSS and JavaScript with the HTML templates. This seems to decouple the build system. I believe this could also aid in creating a single Webpack config to handle all platforms, as there are few differences between the configs. Due to how HtmlWebpackPlugin adds `<link>` tags, I had to add a couple `!important` modifiers in the dev template.

## Assets
Instead of CopyWebpackPlugin, assets (like SVGs) are automatically handled by Webpack's new Asset Module system. This can be extended to dynamic load fonts, like for potential math support.

Webpack now outputs minimized, single bundle files. I analyzed the bundles with various tools thanks to the WebpackConfigDumpPlugin. Finally, I have tested the compiled builds in the Zotero client, and they appear to work well.

I hope some of this exploration may be useful in future work!

Thanks,
Sawyer

### CODE REVIEW

**Comment by mrtcode** [comment]
Great! I'll try it out. Actually, I tried to upgrade Webpack in past, but encountered various random issues, plugin incompatibilities and decided to not waste time anymore. Hopefully we'll succeed this time. And if so, we can then also update pdf-reader, because it has a very similar build system.

**Comment by sawyerpollard** [comment]
> Great! I'll try it out. Actually, I tried to upgrade Webpack in past, but encountered various random issues, plugin incompatibilities and decided to not waste time anymore. Hopefully we'll succeed this time. And if so, we can then also update pdf-reader, because it has a very similar build system.

Sounds great!

File: package.json [hunk 1]
```diff
@@ -3,9 +3,13 @@
...
 	"scripts": {
-		"build": "cross-env NODE_ENV=production webpack --display-modules",
+		"build": "cross-env NODE_ENV=production webpack",
+		"build:dev": "cross-env NODE_ENV=production webpack --config-name dev",
+		"build:web": "cross-env NODE_ENV=production webpack --config-name web",
+		"build:zotero": "cross-env NODE_ENV=production webpack --config-name zotero",
 		"build:ios": "cross-env NODE_ENV=production webpack --config-name ios",
-		"start": "webpack-dev-server",
+		"start": "webpack serve --mode development",
+		"start:dev": "webpack serve --mode development --config-name dev",
```
Comment (mrtcode) [inline package.json:12]: I think we can just have a single `start` option here that would run the `dev` build. Other builds are intended to work only in specific environment therefore running a development server isn't useful with them.

File: webpack.config.js [hunk 2]
```diff
@@ -1,364 +1,288 @@
...
-		publicPath: '/',
-		umdNamedDefine: true
+	name: "zotero",
+	mode: process.env.NODE_ENV === "production" ? "production" : "development",
+	entry: {
+		editor: ["./src/index.zotero.js", "./src/stylesheets/main.scss"],
 	},
 	optimization: {
-		minimize: false
+		minimizer: [new CssMinimizerPlugin(), new TerserPlugin()],
```
Comment (mrtcode) [inline webpack.config.js:153]: We should not minimize code for the Zotero build to have a proper stack trace in Zotero client debug log. Unless we can figure out how Zotero client could use source mapping to report the correct line.

You can see the issue by throwing an error anywhere in note-editor source (i.e. when initialising it).

File: webpack.config.js [hunk 3]
```diff
@@ -1,364 +1,288 @@
...
+const CssMinimizerPlugin = require("css-minimizer-webpack-plugin");
+const TerserPlugin = require("terser-webpack-plugin");
+const { CleanWebpackPlugin } = require("clean-webpack-plugin");
+const HtmlWebpackPlugin = require("html-webpack-plugin");
+const MiniCssExtractPlugin = require("mini-css-extract-plugin");
+const { WebpackConfigDumpPlugin } = require("webpack-config-dump-plugin");
 
 const configDev = {
-	mode: process.env.NODE_ENV === 'production' ? 'production' : 'development',
-	devtool: 'source-map',
```
Comment (mrtcode) [inline webpack.config.js:7]: No source mapping when opening `editor.html` directly.

File: webpack.config.js [hunk 4]
```diff
@@ -1,364 +1,288 @@
...
-			poll: true
-		}
-	}
+	},
 };
 
 const configWeb = {
-	name: 'web',
-	mode: process.env.NODE_ENV === 'production' ? 'production' : 'development',
-	devtool: 'source-map',
```
Comment (mrtcode) [inline webpack.config.js:104]: Source mapping necessary for the web as well.

File: webpack.config.js [hunk 5]
```diff
@@ -1,364 +1,288 @@
...
-		'./src/stylesheets/main.scss'
-	],
+	name: "dev",
+	mode: process.env.NODE_ENV === "production" ? "production" : "development",
+	devtool: process.env.NODE_ENV === "production" ? false : "eval-source-map",
+	entry: {
+		editor: ["./src/index.dev.js", "./src/stylesheets/main.scss"],
+	},
+	optimization: {
+		minimizer: [new CssMinimizerPlugin(), new TerserPlugin()],
```
Comment (mrtcode) [inline webpack.config.js:17]: Probably no need this for the `dev` build.

File: package.json [hunk 6]
```diff
@@ -15,54 +19,56 @@
...
+		"prosemirror-transform": "^1.6.0",
 		"prosemirror-utils": "^0.9.6",
-		"prosemirror-view": "^1.18.8",
+		"prosemirror-view": "^1.26.2",
 		"react": "^16.14.0",
 		"react-dom": "^16.14.0",
-		"react-intl": "^5.20.4"
+		"react-intl": "^6.0.4",
+		"terser-webpack-plugin": "^5.3.3",
+		"webpack-config-dump-plugin": "^3.0.1"
```
Comment (mrtcode) [inline package.json:45]: Packages that are only used for the build process should stay in `devDependencies`.

**Review by mrtcode — COMMENTED** [review]
Aside from those few issues, this update is looking really good.

**Comment by sawyerpollard** [comment]
Thanks for the review items! I think I made most of the changes and a couple extra tweaks. I decided to try manually setting the mode in the webpack config. I'm also trying to use the new webpack `--node-env` flag instead of cross-env, which is in maintenance mode.

**Comment by sawyerpollard** [comment]
I concatenated `process.env.NODE_ENV` to the build filenames, and they appeared to output either 'development' or 'production' based on the `--node-env` flag on Windows 11.

I'm not super well-versed in environment variables and how they're passed to child processes. And if Babel, for example, is a child process that needs NODE_ENV set properly. I'm tempted to imagine that if NODE_ENV is defined in the webpack config, it should be fine.

**Comment by mrtcode** [comment]
> I concatenated `process.env.NODE_ENV` to the build filenames, and they appeared to output either 'development' or 'production' based on the `--node-env` flag on Windows 11.
> 
> I'm not super well-versed in environment variables and how they're passed to child processes. And if Babel, for example, is a child process that needs NODE_ENV set properly. I'm tempted to imagine that if NODE_ENV is defined in the webpack config, it should be fine.

But have you actually tried to build note-editor on Windows? @tnajdek did a good job recently to fix the issues when building on Windows, therefore we have to make sure there aren't any regressions with this PR.

**Comment by tnajdek** [comment]
Guessing from [webpack docs](https://webpack.js.org/configuration/optimization/#optimizationnodeenv) it should work on windows for this particular case but I'd test it just to be sure.

> CSS Tweaks
> 
> Another goal was to get rid of any build errors. I was thankfully able to do this by adding calc() functions where necessary in CSS.

I assume this is to fix an issue with [slash as division](https://sass-lang.com/documentation/breaking-changes/slash-div#automatic-migration)? I think using `math.div` would be more appropriate - it means the final value is calculated by the sass compiler, not the browser.

**Comment by sawyerpollard** [comment]
@tnajdek Thanks for the tips!
I see that Sass says "Alternatively, users can wrap division operations inside a calc() expression, which Sass will simplify to a single number." Just checked and the compiled output doesn't have `calc()` in it, so that looks good I think.

@mrtcode Seems to compile nicely on Windows. Installed Node with NVM.
Screen Shot 2022-07-19 at 5 02 15 PM

I'm not sure if there's any other issues to resolve as of right now. I may see if there are any performance improvements, and I would certainly be interested in exploring similar updates in pdf-editor.

**Comment by tnajdek** [comment]
> I see that Sass says "Alternatively, users can wrap division operations inside a calc() expression, which Sass will simplify to a single number." Just checked and the compiled output doesn't have calc() in it, so that looks good I think.

Ha, that's quite unexpected behavior but good to know that it does simplify some `calc()` calls.

**Comment by sawyerpollard** [comment]
> > I see that Sass says "Alternatively, users can wrap division operations inside a calc() expression, which Sass will simplify to a single number." Just checked and the compiled output doesn't have calc() in it, so that looks good I think.
> 
> Ha, that's quite unexpected behavior but good to know that it does simplify some `calc()` calls.

@tnajdek Do you think I should use `math.div` to be more explicit?

**Comment by tnajdek** [comment]
I would use `math.div` but either one is fine, I just didn't realize sass compiles some `calc()` calls to a value which is an useful but not obvious feature.

**Comment by mrtcode** [comment]
I think we just need to resolve that conflict and we are ready to merge.

This update is really great — not only updates outdated packages with the necessary Webpack config modifications, but also simplifies some things in the build process.

Yes, I think we can do that for pdf-reader as well.

**Comment by sawyerpollard** [comment]
Great! I was able to use math.div instead. I inspected the compiled CSS and it looks good. I see that my branch is a few commits behind master, is this something you will handle or should I?

Thanks!

**Comment by mrtcode** [comment]
Well, there is a conflict, therefore you should resolve it.

**Comment by sawyerpollard** [comment]
> Well, there is a conflict, therefore you should resolve it.

I had to improve by rebasing and file conflict resolution skills, but I think I did everything properly. I will test the build in the Zotero client soon. Additionally, I can improve my commit message or history if needed.

Thanks!

**Comment by mrtcode** [comment]
Could you resolve the conflict once again. Sorry for that. We'll merge it afterwards.

**Comment by sawyerpollard** [comment]
@mrtcode I made a few tweaks to get it working with math. Also, after testing the build in Zotero, I changed a few Webpack settings. There are a few errors related to note-editor in Zotero client, but not noticeably more than in the default Zotero client.

Final note, in the dropdown toolbar, the insert math button says "noteEditor.mathBlock". I assume this is because localization hasn't been added to Zotero client.

I'll leave this script I found very helpful for testing here too (on Mac).
```bash
#!/bin/bash

# Setup
ZOTERO_DIR=zotero-client
ZOTERO_BUILD_DIR=zotero-standalone-build
NOTE_EDITOR_DIR=note-editor
pkill -x zotero

# Build Zotero
cd $ZOTERO_DIR
npm run clean-build
cd ..

# Build Note Editor
cd $NOTE_EDITOR_DIR
npm run build:zotero
rsync -a ./build/zotero/* "../$ZOTERO_DIR/build/resource/note-editor/"
cd ..

# Package App
cd $ZOTERO_BUILD_DIR
./scripts/dir_build
open ./staging/Zotero.app
cd ..
```

**Comment by mrtcode** [comment]
Does the latest commit intentionally remove KaTeX fonts (and copies them dynamically?) or things just messed up when rebasing? Anything else is missing?

**Comment by mrtcode** [comment]
Ok, I see now that it takes font files directly from KaTeX package. Great! Sorry, ignore my last comment.

> Final note, in the dropdown toolbar, the insert math button says "noteEditor.mathBlock". I assume this is because localization hasn't been added to Zotero client.

Yes, we'll need to add the string somewhere [here](https://github.com/zotero/zotero/blob/6921268dc0356474ca33ff732e3aa80a004cbc7c/chrome/locale/en-US/zotero/zotero.properties#L1301) at the time when we update note-editor submodule in zotero repo.

**Comment by sawyerpollard** [comment]
Great, thank you!

**Pull request merged**